### DISCUSSION TITLE
Trying to get Typescript working

### DISCUSSION THREAD

**Discussion by ktwbc** [tywalch/electrodb#282] [Q&A]
I've gone through the docs on trying to use the Typescript types correctly but can't seem to get it working. I have two issues.

First, I just have an extremely simple helper function where I can pass parameters and I'm trying to get the return type right. I have:

```
export const AuditTrail = new Entity(
  {
    model: {
      entity: 'AuditTrail',
   ...etc
    attributes: {
      user: { type: 'string' },
      token: { type: 'string', required: false },
    etc...
}

type EntitySchema = CreatedItem<typeof AuditTrail>;

export const write = (
  user: string,
  event: AuditLogEvent,
  token?: string,
  data?: object
): Promise<ResponseItem<typeof AuditTrail>> => {
  return AuditTrail.create({ data, event, token, user }).go();
};
```

I've tried all variations but typescript (5.1.6) is flatting it no matter what.

The second issue is that on my write() function I have two required and two optional parameters. Since they may be undefind, I should be able to include them verbatim and with data this works, but it's complaining that token can be undefined which is invalid, even though in the attributes it's set to required: false 
<img width="1341" alt="Screenshot 2023-08-19 at 11 21 58" src="https://github.com/tywalch/electrodb/assets/624611/c45b66d6-ec82-489d-9c14-5e26d0f54f3a">

**Comment by ktwbc** [comment]
This appears to be related to the fact that my token item is used in a GSI but it's an optional field (meaning the GSI would simply not contain that entry) but electro is complaining that because it's used in a GSI it's required anyway, which is not true according to dynamo

**Comment by ktwbc** [comment]
so further on GSI issue, I have my indexes (primary and two GSIs setup this way)

```
 indexes: {
      byUser: {
        pk: {
          field: 'user',
          composite: ['user'],
          template: '${user}'
        },
        sk: {
          field: 'created_at',
          composite: ['created_at']
        }
      },
      byEvent: {
        index: 'event-created',
        pk: {
          field: 'event',
          composite: ['event'],
          template: '${event}'
        },
        sk: {
          field: 'created_at',
          composite: ['created_at']
        }
      },
      byToken: {
        index: 'token-created',
        pk: {
          field: 'token',
          composite: ['token'],
         template: '${token}'
        },
        sk: {
          field: 'created_at',
          composite: ['created_at']
        }
      }
```
but not all my records have token and that's okay for a GSI but not for electro. if I comment out the byToken block my insert works, but then I can't query using the GSI via electro. I guess I can make a second Entity for the same table, is that the best option?

**Reply by tywalch** [reply]
Are you able to recreate this using the playground: https://electrodb.fun
